File: markdown/self-drive-car/src/radar.ts
import { drawLine, drawText, lineInfo, posInfo } from "./drawUtils";
import { CarHeight, CarWidth, RadarColor, RadarLineLength } from "./gameUtils";
import { EmptyPosition, ILine, IPosition, findTwoLinesIntersection, getDistance, getLineSlope, getTwoPointsDistance, rotatePoints } from "./math";

export class RadarLine {
    pos: IPosition = EmptyPosition;
    carXY: IPosition = EmptyPosition;
    angle: number = 0;

    render(ctx: CanvasRenderingContext2D) {
        const { start, end } = this.getDrawLine();
        const [start1, end1] = rotatePoints(this.pos, this.angle, [start, end]);
        drawLine(ctx, { start: start1, end: end1 }, { strokeSyle: RadarColor, lineWidth: 3 });

        //draw dump info
        //const radarLine = this.getBoundLine();
        //drawText(ctx, start1, `${posInfo(radarLine.start)}`)
        //drawText(ctx, end1, `${posInfo(radarLine.end)}`)
    }

    renderDamaged(ctx: CanvasRenderingContext2D, point: IPosition) {
        const { start, end } = this.getBoundLine();
        const distance = getTwoPointsDistance(start, point);
        const { start: startDrawPos, end: endDrawPos } = this.getDrawLine();
        const endPos = {
            x: startDrawPos.x,
            y: startDrawPos.y - distance,
        };
        const [startPos1, endPos1] = rotatePoints(this.pos, this.angle, [startDrawPos, endPos]);
        const radarLine = {
            start: startPos1,
            end: endPos1,
        };
        drawLine(ctx, radarLine, { strokeSyle: 'yellow', lineWidth: 3 });
    }
    
    getDrawLine() {
        const start = {
            x: this.pos.x,
            y: this.pos.y - CarHeight / 2 + 20,
        };
        const end = {
            x: start.x,
            y: start.y - RadarLineLength
        };
        return { start, end };
    }
    
    getBoundLine() {
        const start = {
            x: this.carXY.x,
            y: this.carXY.y - CarHeight / 2 + 20,
        };
        const end = {
            x: start.x,
            y: start.y - RadarLineLength
        };
        const [start1, end1] = rotatePoints(this.carXY, this.angle, [start, end]);
        return { start: start1, end: end1 };
    }
    
    isMy(point: IPosition): boolean {
        const { start, end } = this.getBoundLine();
        const slope = getLineSlope({ start, end });
        const testLine = { start, end: point };
        const testSlope = getLineSlope(testLine);
        if( slope === testSlope ) {
            return true;
        }
        return false;
    }
}

export class Radar {
    pos: IPosition = { x: 0, y: 0 };
    angle = 270;
    carXY: IPosition = { x: 0, y: 0 };

    radarLine: ILine | null = null;

    render(ctx: CanvasRenderingContext2D) {
        const { start, end } = this.getDrawLine();
        const [start1, end1] = rotatePoints(this.pos, this.angle, [start, end]);
        drawLine(ctx, { start: start1, end: end1 }, { strokeSyle: RadarColor, lineWidth: 3 });

        const radarLine = this.getBoundLines()[0];
        drawText(ctx, start1, `${posInfo(radarLine.start)}`)
        drawText(ctx, end1, `${posInfo(radarLine.end)}`)
    }

    getBoundLines() {
        const start = {
            x: this.carXY.x,
            y: this.carXY.y - CarHeight / 2 + 20,
        };
        const end = {
            x: start.x,
            y: start.y - RadarLineLength
        };
        const [start1, end1] = rotatePoints(this.carXY, this.angle, [start, end]);
        return [{ start: start1, end: end1 }];
    }

    renderDamaged(ctx: CanvasRenderingContext2D, point: IPosition) {
        const { start, end } = this.getBoundLines()[0];
        const distance = getTwoPointsDistance(start, point);
        const { start: startDrawPos, end: endDrawPos } = this.getDrawLine();
        const endPos = {
            x: startDrawPos.x,
            y: startDrawPos.y - distance,
        };
        const [startPos1, endPos1] = rotatePoints(this.pos, this.angle, [startDrawPos, endPos]);
        const radarLine = {
            start: startPos1,
            end: endPos1,
        };
        drawLine(ctx, radarLine, { strokeSyle: 'yellow', lineWidth: 3 });
    }

    getDrawLine() {
        const start = {
            x: this.pos.x,
            y: this.pos.y - CarHeight / 2 + 20,
        };
        const end = {
            x: start.x,
            y: start.y - RadarLineLength
        };
        return { start, end };
    }
}

File: torch-self-drive-car/radar.py
import math

from game import RadarColor, CarHeight, RadarLineLength, RadarLineCount
from math_utils import Position, rotate_points, Line
from pygameGraphic import IGraphic
from roads import RoadMap


class RadarLine:
    def __init__(self):
        self.pos = Position(0, 0)
        self.car_xy = Position(0, 0)
        self.car_angle = 0
        self.angle = 0
        self.distance = 0

    def render(self, ctx: IGraphic):
        ctx.draw_line(self.get_draw_line(True), color=RadarColor, thickness=3)

    def render_damaged(self, ctx: IGraphic, point: Position):
        start, end = self.get_bound_line()
        distance = round(math.sqrt((start.x - point.x) ** 2 + (start.y - point.y) ** 2))
        self.distance = distance
        draw_line = self.get_draw_line(False)
        end_pos = Position(draw_line.start.x, draw_line.start.y + distance)
        rotated_points = rotate_points(self.pos, self.car_angle + self.angle, [draw_line.start, end_pos])
        radar_line = Line(rotated_points[0], rotated_points[1])
        ctx.draw_line(radar_line, color="red", thickness=5)

    def get_draw_line(self, rotate: bool) -> Line:
        start = Position(self.pos.x, self.pos.y + CarHeight / 2 - 20)
        end = Position(start.x, start.y + RadarLineLength)
        if rotate:
            rotated_points = rotate_points(self.pos, self.car_angle + self.angle, [start, end])
            return Line(rotated_points[0], rotated_points[1])
        return Line(start, end)

    def get_bound_line(self) -> Line:
        start = Position(self.car_xy.x, self.car_xy.y + CarHeight / 2 - 20)
        end = Position(start.x, start.y + RadarLineLength)
        rotated_points = rotate_points(self.car_xy, self.car_angle + self.angle, [start, end])
        return Line(rotated_points[0], rotated_points[1])

    def is_my(self, point: Position) -> bool:
        start, end = self.get_bound_line()
        slope = (end.y - start.y) / (end.x - start.x)
        test_line = Line(start, point)
        test_slope = (test_line.end.y - test_line.start.y) / (test_line.end.x - test_line.start.x)
        if self.compare_with_tolerance(slope, test_slope):
            return True
        self.distance = RadarLineLength
        return False

    @staticmethod
    def compare_with_tolerance(a, b):
        return abs(a - b) < 0.0000001


def generate_angles(count: int, angle: int) -> list[int]:
    angles = []
    start_angle = 0
    counter = 0
    if count % 2 != 0:
        angles.append(start_angle)
        counter = 1

    for i in range(1, count):
        if i % 2 == 1:
            angles.append(start_angle - counter * angle)
        else:
            angles.append(start_angle + counter * angle)
            counter += 1

    return angles


class Radar:
    radar_lines: list[RadarLine] = []

    def __init__(self):
        self.pos = Position(0, 0)
        self.car_angle = 0
        self.car_xy = Position(0, 0)
        for radar_angle in generate_angles(RadarLineCount, 30):
            radar_line = RadarLine()
            radar_line.angle = radar_angle
            self.radar_lines.append(radar_line)

    def render(self, ctx: IGraphic):
        for radar_line in self.radar_lines:
            radar_line.pos = self.pos
            radar_line.car_xy = self.car_xy
            radar_line.car_angle = self.car_angle
            radar_line.render(ctx)

    def collide(self, ctx: IGraphic, road_map: RoadMap):
        for index, radar_line in enumerate(self.get_bound_lines()):
            radar = self.radar_lines[index]
            road, collide_radar_points = road_map.collide(ctx, [radar_line])
            if collide_radar_points:
                radar.render_damaged(ctx, collide_radar_points[0])
            else:
                radar.distance = RadarLineLength

    def get_bound_lines(self) -> list[Line]:
        bound_lines = []
        for radar_line in self.radar_lines:
            radar_line.pos = self.pos
            radar_line.car_xy = self.car_xy
            radar_line.car_angle = self.car_angle
            bound_lines.append(radar_line.get_bound_line())
        return bound_lines

    def render_damaged(self, ctx: IGraphic, point: Position):
        for radar_line in self.radar_lines:
            radar_line.pos = self.pos
            radar_line.car_xy = self.car_xy
            radar_line.car_angle = self.car_angle
            if radar_line.is_my(point):
                radar_line.render_damaged(ctx, point)
